File: backend/src/merchant/api/MerchantCreateProductApi.scala
package delivery.merchant.api

import cats.effect.IO
import delivery.merchant.objects.{CreateProductRequest, CreateProductResponse}
import delivery.merchant.state.MerchantDomainOps
import delivery.shared.api.ApiPlan
import delivery.shared.objects.DeliveryState
import delivery.shared.state.DeliveryStateOps
import org.typelevel.log4cats.slf4j.Slf4jLogger

object MerchantCreateProductApi extends ApiPlan[
  MerchantCreateProductApi.MerchantCreateProductCommand,
  Either[String, MerchantCreateProductApi.MerchantCreateProductSuccess]
]:

  final case class MerchantCreateProductCommand(
      state: DeliveryState,
      username: String,
      body: CreateProductRequest
  )

  final case class MerchantCreateProductSuccess(nextState: DeliveryState, response: CreateProductResponse)

  private val logger = Slf4jLogger.getLogger[IO]

  override val name: String = "MerchantCreateProductApi"

  override def plan(input: MerchantCreateProductCommand): IO[Either[String, MerchantCreateProductSuccess]] =
    for
      _ <- logger.info(s"$name started, username=${input.username}, merchantId=${input.body.merchantId}")
      response <- MerchantDomainOps.createProduct(
        input.state.merchant,
        input.username,
        input.body.merchantId,
        input.body.name,
        input.body.description,
        input.body.price,
        input.body.remainingStock,
        input.body.listingStatus
      ).map(_.map { case (nextMerchant, createdProduct) =>
        MerchantCreateProductSuccess(
          DeliveryStateOps.withMerchantState(input.state, nextMerchant),
          CreateProductResponse(ok = true, product = createdProduct)
        )
      })
      _ <- logger.info(s"$name finished, success=${response.isRight}")
    yield response

end MerchantCreateProductApi

File: frontend/src/api/merchant/MerchantCreateProductApi.ts
import { APIMessage } from '@/api/shared/APIMessage'
import type { TaskIO } from '@/api/shared/TaskIO'
import { sendAPI } from '@/api/shared/sendAPI'
import type { CreateProductRequest } from '@/objects/merchant/CreateProductRequest'
import type { Product } from '@/objects/merchant/Product'
import type { ListingStatus, MerchantId } from '@/objects/shared/ids'

class MerchantCreateProductAPI extends APIMessage<Product> {
  readonly apiName = 'merchantcreateproductapi'
  readonly merchantId: MerchantId
  readonly name: string
  readonly description: string
  readonly price: number
  readonly remainingStock: number
  readonly listingStatus: ListingStatus

  constructor(
    merchantId: MerchantId,
    name: string,
    description: string,
    price: number,
    remainingStock: number,
    listingStatus: ListingStatus,
  ) {
    super()
    this.merchantId = merchantId
    this.name = name
    this.description = description
    this.price = price
    this.remainingStock = remainingStock
    this.listingStatus = listingStatus
  }
}

export function createMerchantProductIO(input: CreateProductRequest): TaskIO<Product> {
  return sendAPI(
    new MerchantCreateProductAPI(
      input.merchantId,
      input.name,
      input.description,
      input.price,
      input.remainingStock,
      input.listingStatus,
    ),
  )
}
